### PR TITLE
Add architecture document

### PR DESCRIPTION
**Pull request by kurapov-peter**

### CODE REVIEW

File: docs/ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+
+![](pics/pt2.png)
+
+From pytorch point of view, a model FX graph [3] is captured by Dynamo [4] (and AOT Autograd if necessary) which is passed to Inductor [5]. The Inductor converts the broad set of pytorch’s ATen operators into a narrower set of stable primitive operators (PrimTorch) [6] via “decompositions” (a customizable mechanism for complex operator expansion). The resulting graph is translated into Triton kernels which are decorated python functions produced by the Inductor [7]. Triton frontend consumes function sources, the internals follow traditional compiler architecture: an AST is built and traversed to produce an MLIR module. The module is progressively lowered into the LLVM dialect. A regular LLVM module is then produced and converted to an appropriate representation depending on the target (e.g., PTX for Nvidia). A compiled kernel is then returned with a launcher stub. Kernels are then executed by pytorch through Inductor’s scheduling & wrapper interfaces.
```
Comment (aregm) [inline docs/ARCHITECTURE.md:11]: /s/pytorch/PyTorch

File: docs/ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+    x2 = xindex
+    tmp0 = tl.load(in_ptr0 + (x0), xmask, eviction_policy='evict_last')
+    tmp1 = tl.load(in_out_ptr0 + (x2), xmask)
+    tmp2 = tmp0 + tmp1
+    tmp3 = triton_helpers.maximum(0, tmp2)
+    tl.store(in_out_ptr0 + (x2), tmp3, xmask)
+```
+
+Each kernel will additionally be annotated with meta information such as size hints, signature, device type, mutated args, and others.
+By default, Inductor reuses efficient implementations of compute-heavy operations such as matrix multiplication or flash attention via dispatching a call to the Aten op native implementation. The generated code for a single causal attention block would look like the following. Note the linear layers and the attention blocks produce a standalone kernel call (only the fused native dropout makes it to a Triton kernel in this case).
```
Comment (aregm) [inline docs/ARCHITECTURE.md:53]: ...call to the operator's native implementation in Aten.

File: docs/ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+
```
Comment (aregm) [inline docs/ARCHITECTURE.md:54]: What is the original code to see before/after?

File: docs/ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+    row_minus_max = row - tl.max(row, axis=0)
+    numerator = tl.exp(row_minus_max)
+    denominator = tl.sum(numerator, axis=0)
+    softmax_output = numerator / denominator
+    output_row_start_ptr = output_ptr + row_idx * output_row_stride
+    output_ptrs = output_row_start_ptr + col_offsets
+    tl.store(output_ptrs, softmax_output, mask=col_offsets< n_cols)
+```
+
+The resulting IR follows the input language almost 1 to 1:
```
Comment (aregm) [inline docs/ARCHITECTURE.md:120]: This is good.

File: docs/ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+* **Graph operators’ decomposition** into a smaller set of operators (~250 PrimTorch ops instead of ~2k pytorch ops [11]; Note: the sets are not stabilized!).
+* **Loop-level IR graph lowering**: remove views, broadcasting, indexing simplification, materialization vs reuse decisions, layout optimization, loop reordering.
+* **Scheduling**: horizontal/vertical/reduction fusion , tiling, memory planning, buffer reuse, autotuning.
+* **Backend code generation** (depending on the target).
+* **Wrapper code generation**: basically, an unrolled interpreter loop to run the generated kernel (gather the inputs, symbolic shape processing, allocate the tensors, and invoke the kernel).
+The sequence of the above steps produces a `call` function that controls the guards [13] and then calls a sequence of Aten/Extern kernels intermixed with generated triton functions. Former are resolved through an algorithm selection mechanism [12]. The latter is a Python function with `@triton.jit` decorator, e.g. (fused bias from a linear layer and a Relu):
+
+```python
```
Comment (aregm) [inline docs/ARCHITECTURE.md:36]: import torch
import triton
import triton.language as tl

File: docs/ARCHITECTURE.md [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+* Tensor pointer rewriting.
+
+#### TritonGPU dialect
+TritonGPU dialect [18] exposes GPU-specific operators. After converting Trition dialect to TritonGPU the following set of optimizations are run:
+* Coalescing – make sure the dimension with greatest contiguity is first.
+* Layout conversion removal.
+* Thread locality optimization.
+* Matmul acceleration pipeline.
+* Dot operands optimization.
+* SW Pipelining.
```
Comment (aregm) [inline docs/ARCHITECTURE.md:175]: What is SW?

**Review by aregm — APPROVED** [review]
With the updates.

File: docs/ARCHITECTURE.md (hunk 3 above)
Comment (kurapov-peter) [reply docs/ARCHITECTURE.md:115]: Added the model source code

File: docs/ARCHITECTURE.md (hunk 5 above)
Comment (kurapov-peter) [reply docs/ARCHITECTURE.md:36]: Added

**Comment by kurapov-peter** [comment]
@pbchekin, could we disable the default CI pipeline for changes to the documentation?
